File: src/main/remoteBackupStore.ts
/**
 * Cloudflare R2 remote backup (S3-compatible API).
 * Stores `bundle.json.gz` + `manifest.json` under a configurable prefix.
 */

import {
  GetObjectCommand,
  HeadBucketCommand,
  HeadObjectCommand,
  PutObjectCommand,
  S3Client,
} from "@aws-sdk/client-s3";
import type { Readable } from "stream";

export const BUNDLE_OBJECT_NAME = "bundle.json.gz";
export const MANIFEST_OBJECT_NAME = "manifest.json";

export interface BackupManifest {
  version: number;
  revision: string;
  contentRevision?: string;
  updatedAt: number;
  bundleHash: string;
}

export interface R2Config {
  accountId: string;
  bucket: string;
  prefix: string;
  accessKeyId: string;
  secretAccessKey: string;
}

export function normalizeR2Prefix(prefix: string): string {
  const trimmed = prefix.trim().replace(/^\/+/, "");
  if (!trimmed) return "harness/";
  return trimmed.endsWith("/") ? trimmed : `${trimmed}/`;
}

export function r2Endpoint(accountId: string): string {
  return `https://${accountId.trim()}.r2.cloudflarestorage.com`;
}

function objectKey(prefix: string, name: string): string {
  return `${normalizeR2Prefix(prefix)}${name}`;
}

async function streamToBuffer(body: Readable | ReadableStream | Blob | undefined): Promise<Buffer> {
  if (!body) return Buffer.alloc(0);
  if (body instanceof Blob) {
    return Buffer.from(await body.arrayBuffer());
  }
  if (typeof (body as ReadableStream).getReader === "function") {
    const reader = (body as ReadableStream<Uint8Array>).getReader();
    const chunks: Uint8Array[] = [];
    for (;;) {
      const { done, value } = await reader.read();
      if (done) break;
      if (value) chunks.push(value);
    }
    return Buffer.concat(chunks.map((c) => Buffer.from(c)));
  }
  const nodeStream = body as Readable;
  const chunks: Buffer[] = [];
  for await (const chunk of nodeStream) {
    chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk));
  }
  return Buffer.concat(chunks);
}

export class RemoteBackupStore {
  private readonly client: S3Client;
  private readonly bucket: string;
  private readonly prefix: string;

  constructor(config: R2Config) {
    this.bucket = config.bucket.trim();
    this.prefix = normalizeR2Prefix(config.prefix);
    this.client = new S3Client({
      region: "auto",
      endpoint: r2Endpoint(config.accountId),
      credentials: {
        accessKeyId: config.accessKeyId.trim(),
        secretAccessKey: config.secretAccessKey,
      },
    });
  }

  manifestKey(): string {
    return objectKey(this.prefix, MANIFEST_OBJECT_NAME);
  }

  bundleKey(): string {
    return objectKey(this.prefix, BUNDLE_OBJECT_NAME);
  }

  async readManifest(): Promise<BackupManifest | null> {
    try {
      const response = await this.client.send(
        new GetObjectCommand({
          Bucket: this.bucket,
          Key: this.manifestKey(),
        }),
      );
      const bytes = await streamToBuffer(response.Body as Readable);
      const raw = JSON.parse(bytes.toString("utf-8")) as Partial<BackupManifest>;
      if (
        typeof raw.revision === "string" &&
        typeof raw.updatedAt === "number" &&
        typeof raw.bundleHash === "string" &&
        typeof raw.version === "number"
      ) {
        return {
          version: raw.version,
          revision: raw.revision,
          contentRevision: typeof raw.contentRevision === "string" ? raw.contentRevision : undefined,
          updatedAt: raw.updatedAt,
          bundleHash: raw.bundleHash,
        };
      }
      return null;
    } catch (err: unknown) {
      const name = (err as { name?: string })?.name;
      const status = (err as { $metadata?: { httpStatusCode?: number } })?.$metadata?.httpStatusCode;
      if (name === "NoSuchKey" || status === 404) return null;
      throw err;
    }
  }

  async readBundle(): Promise<Buffer> {
    const response = await this.client.send(
      new GetObjectCommand({
        Bucket: this.bucket,
        Key: this.bundleKey(),
      }),
    );
    return streamToBuffer(response.Body as Readable);
  }

  async writeBundleAndManifest(params: {
    bundleBytes: Buffer;
    manifest: BackupManifest;
  }): Promise<void> {
    const { bundleBytes, manifest } = params;
    await this.client.send(
      new PutObjectCommand({
        Bucket: this.bucket,
        Key: this.bundleKey(),
        Body: bundleBytes,
        ContentType: "application/gzip",
      }),
    );
    await this.client.send(
      new PutObjectCommand({
        Bucket: this.bucket,
        Key: this.manifestKey(),
        Body: Buffer.from(JSON.stringify(manifest, null, 2), "utf-8"),
        ContentType: "application/json",
      }),
    );
  }

  async testConnection(): Promise<{ ok: true } | { ok: false; error: string }> {
    try {
      await this.client.send(new HeadBucketCommand({ Bucket: this.bucket }));
      try {
        await this.client.send(
          new HeadObjectCommand({ Bucket: this.bucket, Key: this.manifestKey() }),
        );
      } catch (err: unknown) {
        const status = (err as { $metadata?: { httpStatusCode?: number } })?.$metadata?.httpStatusCode;
        if (status !== 404) throw err;
      }
      return { ok: true };
    } catch (err) {
      return {
        ok: false,
        error: err instanceof Error ? err.message : String(err),
      };
    }
  }
}

export function isR2ConfigComplete(
  sync: { accountId?: string; bucket?: string; prefix?: string; accessKeyId?: string } | undefined,
  hasSecret: boolean,
): boolean {
  if (!sync || !hasSecret) return false;
  return Boolean(
    sync.accountId?.trim() &&
      sync.bucket?.trim() &&
      sync.accessKeyId?.trim(),
  );
}

File: ios/HarnessMobile/Sync/RemoteBackupStore.swift
import AsyncHTTPClient
import Foundation
import NIOCore
import SotoCore
import SotoS3

struct R2Config: Sendable {
    let accountId: String
    let bucket: String
    let prefix: String
    let accessKeyId: String
    let secretAccessKey: String
}

enum RemoteBackupStoreError: LocalizedError {
    case notConfigured
    case invalidManifest

    var errorDescription: String? {
        switch self {
        case .notConfigured:
            return "Cloudflare R2 is not configured."
        case .invalidManifest:
            return "Remote backup manifest is invalid."
        }
    }
}

/// Cloudflare R2 remote backup (S3-compatible API).
/// Stores content-addressed `bundle-<hash>.json.gz` + `manifest.json` under a configurable prefix,
/// and mirrors bytes to legacy `bundle.json.gz` for older clients.
final class RemoteBackupStore: @unchecked Sendable {
    private let s3: S3
    private let bucket: String
    private let prefix: String
    private let awsClient: AWSClient
    /// Owned client with HTTP decompression disabled.
    /// `HTTPClient.shared` enables gzip and fails on precompressed `bundle.json.gz` from R2
    /// (`NIOHTTPDecompression.ExtraDecompressionError`).
    private let httpClient: HTTPClient

    init(config: R2Config) {
        bucket = config.bucket.trimmingCharacters(in: .whitespacesAndNewlines)
        prefix = R2SettingsStore.normalizePrefix(config.prefix)
        httpClient = HTTPClient(
            eventLoopGroupProvider: .singleton,
            configuration: HTTPClient.Configuration(decompression: .disabled)
        )
        awsClient = AWSClient(
            credentialProvider: .static(
                accessKeyId: config.accessKeyId.trimmingCharacters(in: .whitespacesAndNewlines),
                secretAccessKey: config.secretAccessKey
            ),
            httpClient: httpClient
        )
        let endpoint = Self.r2Endpoint(accountId: config.accountId)
        s3 = S3(
            client: awsClient,
            region: .other("auto"),
            endpoint: endpoint,
            options: [.s3DisableChunkedUploads]
        )
    }

    deinit {
        try? awsClient.syncShutdown()
        try? httpClient.syncShutdown()
    }

    static func r2Endpoint(accountId: String) -> String {
        "https://\(accountId.trimmingCharacters(in: .whitespacesAndNewlines)).r2.cloudflarestorage.com"
    }

    static func objectKey(prefix: String, name: String) -> String {
        "\(R2SettingsStore.normalizePrefix(prefix))\(name)"
    }

    static func contentAddressedBundleObjectName(bundleHash: String) -> String {
        "bundle-\(bundleHash).json.gz"
    }

    static func makeConfigured() throws -> RemoteBackupStore {
        guard R2SettingsStore.isConfigured,
              let secret = KeychainStore.loadR2SecretAccessKey() else {
            throw RemoteBackupStoreError.notConfigured
        }
        return RemoteBackupStore(config: R2Config(
            accountId: R2SettingsStore.accountId,
            bucket: R2SettingsStore.bucket,
            prefix: R2SettingsStore.prefix,
            accessKeyId: R2SettingsStore.accessKeyId,
            secretAccessKey: secret
        ))
    }

    func manifestKey() -> String {
        Self.objectKey(prefix: prefix, name: SyncScopes.manifestFileName)
    }

    func legacyBundleKey() -> String {
        Self.objectKey(prefix: prefix, name: SyncScopes.legacyBundleFileName)
    }

    func bundleKey(forHash bundleHash: String) -> String {
        Self.objectKey(prefix: prefix, name: Self.contentAddressedBundleObjectName(bundleHash: bundleHash))
    }

    /// Legacy fixed key — prefer `bundleKey(forHash:)`.
    func bundleKey() -> String {
        legacyBundleKey()
    }

    func readManifest() async throws -> BackupManifest? {
        do {
            let response = try await s3.getObject(.init(bucket: bucket, key: manifestKey()))
            let bytes = try await response.body.collect(upTo: 1_048_576)
            return BackupManifest.load(fromData: Data(bytes.readableBytesView))
        } catch {
            if isNotFound(error) { return nil }
            throw error
        }
    }

    /// Read the bundle for `bundleHash`, preferring the content-addressed object and
    /// falling back to the legacy fixed key for manifests written by older clients.
    func readBundle(bundleHash: String) async throws -> Data {
        do {
            let response = try await s3.getObject(.init(bucket: bucket, key: bundleKey(forHash: bundleHash)))
            let bytes = try await response.body.collect(upTo: 64 * 1024 * 1024)
            return Data(bytes.readableBytesView)
        } catch {
            if !isNotFound(error) { throw error }
            let response = try await s3.getObject(.init(bucket: bucket, key: legacyBundleKey()))
            let bytes = try await response.body.collect(upTo: 64 * 1024 * 1024)
            return Data(bytes.readableBytesView)
        }
    }

    func writeBundleAndManifest(
        bundleBytes: Data,
        manifest: BackupManifest,
        previousBundleHash: String? = nil
    ) async throws {
        let addressedKey = bundleKey(forHash: manifest.bundleHash)
        _ = try await s3.putObject(.init(
            body: AWSHTTPBody(bytes: bundleBytes),
            bucket: bucket,
            contentType: "application/gzip",
            key: addressedKey
        ))

        // Compat mirror for clients that still read the fixed key.
        _ = try? await s3.putObject(.init(
            body: AWSHTTPBody(bytes: bundleBytes),
            bucket: bucket,
            contentType: "application/gzip",
            key: legacyBundleKey()
        ))

        let manifestData = try JSONEncoder().encode(manifest)
        _ = try await s3.putObject(.init(
            body: AWSHTTPBody(bytes: manifestData),
            bucket: bucket,
            contentType: "application/json",
            key: manifestKey()
        ))

        if let previousBundleHash,
           !previousBundleHash.isEmpty,
           previousBundleHash != manifest.bundleHash {
            _ = try? await s3.deleteObject(.init(
                bucket: bucket,
                key: bundleKey(forHash: previousBundleHash)
            ))
        }
    }

    func testConnection() async -> (ok: Bool, error: String?) {
        do {
            _ = try await s3.headBucket(.init(bucket: bucket))
            do {
                _ = try await s3.headObject(.init(bucket: bucket, key: manifestKey()))
            } catch {
                if !isNotFound(error) { throw error }
            }
            return (true, nil)
        } catch {
            return (false, error.localizedDescription)
        }
    }

    private func isNotFound(_ error: Error) -> Bool {
        if let awsError = error as? AWSResponseError {
            if awsError.errorCode.lowercased().contains("nosuchkey") { return true }
            if awsError.context?.responseCode == .notFound { return true }
        }
        if let rawError = error as? AWSRawError, rawError.context.responseCode == .notFound {
            return true
        }
        let message = error.localizedDescription.lowercased()
        return message.contains("404") || message.contains("not found") || message.contains("nosuchkey")
    }
}
